Initial Implementation of Autocomplete

dnachum opened the pull request:
Closes #74 

Initially the autocomplete feature is off until it is toggled with (autocomplete) but the window should save whether autocomplete was toggled so students won't have to keep enabling/disabling it.

The way I intended for this to work was for words to be displayed that the user could be typing. Once there is only one possible word left, the docs for that procedure is displayed. Also, for longer words, you can press tab to autocomplete the word that was being typed.

There are some things that can definitely be improved from this point. One thing is to maybe make the matching more sophisticated (right now I only match the prefixes of the word that is being typed with the possible words that can be displayed). Another thing is for something like define to not have the autocomplete box disappear when typing the function name and arguments.

jathak commented:
Great work on this! I started my new job yesterday, so my time's more limited now, but I'll try to get this reviewed this weekend

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -101,6 +121,43 @@ class CodeInput {
   }
 
+  /// Determines the operation at the last open parens
+  List _currOp(String inputText, int position, [int fromLast = 1]) {
```
It's usually best not to use a list to return multiple values of different types, since we lose type checking.

I'm okay with it if we include a detailed description of the two values (including types) in the doc comment, but I think the better way to go would be to use something like the [tuple package](https://pub.dartlang.org/packages/tuple)

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -145,8 +202,75 @@ class CodeInput {
+    for (String schemeWord in wordToDocs.keys) {
+      if (((schemeWord.length > currWord.length) && !fullWord) ||
+          schemeWord.length == currWord.length) {
+        if (schemeWord.substring(0, currLength) == currWord) {
```
The declared name in line 208 isn't really much shorter and doesn't save any repeated work (plus we already use `currWord.length` in the two lines above), so just replace `currLength` with `currWord.length` here and eliminate the declaration.

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -101,6 +121,43 @@ class CodeInput {
+        } else if (nextOpen == -1 || nextClose == -1 || nextOpen < nextClose) {
+          //Assuming the word is right after the parens
+          List splitRef =
+              refLine.substring(strIndex + 1).split(RegExp("[ ()]+"));
```
Not sure what this regular expression is actually doing. I recommend including a comment describing the meaning whenever you use all but the most trivial regexes, since they are notoriously difficult to understand.

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -145,8 +202,75 @@ class CodeInput {
+    //Find the text to the left of where the typing cursor currently is
+    int cursorPos = findPosition(element, window.getSelection().getRangeAt(0));
+    List<String> inputText =
+        element.text.substring(0, cursorPos).split(RegExp("[(]+"));
```
This regex should probably be commented too

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -145,8 +202,75 @@ class CodeInput {
+    //Find the last word that was being typed
+    int currLength = 0;
+    List output = _currOp(element.text, cursorPos);
+    List output2 = _currOp(element.text, cursorPos, 2);
```
Can you add a comment explaining the difference between `output` and `output2`?

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -145,8 +202,75 @@ class CodeInput {
+    int currLength = 0;
+    List output = _currOp(element.text, cursorPos);
+    List output2 = _currOp(element.text, cursorPos, 2);
+    String findMatch = output[0];
```
Noun phrases are [generally preferred](https://www.dartlang.org/guides/language/effective-dart/design#prefer-a-noun-phrase-for-a-non-boolean-property-or-variable) for non-boolean variables.
A better name would perhaps be `foundMatch` or just `match`

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -145,8 +202,75 @@ class CodeInput {
+    List output = _currOp(element.text, cursorPos);
+    List output2 = _currOp(element.text, cursorPos, 2);
+    String findMatch = output[0];
+    bool fullWord = output[1];
```
Likewise, Dart [prefers](https://www.dartlang.org/guides/language/effective-dart/design#prefer-a-non-imperative-verb-phrase-for-a-boolean-property-or-variable) that booleans are named with a form of "to be" most of the time.

`isFullWord` would probably be better here.

dnachum commented:
Thanks for reviewing this! I'll be sure to get working on these things this week and I'll also read through Effective Dart.

dnachum commented:
Sorry it took me so long to get to this. I think I fixed most of the styling issues and also added a blurb about it in the "Other Useful Commands" section.

jathak commented on lib/src/web_ui/code_input.dart:
```diff
@@ -14,22 +17,35 @@ const List<SchemeSymbol> noIndentForms = [
   SchemeSymbol("define-macro")
 ];
 
+bool _isEnabledAutocomplete = false;
```
Nit: `_isAutocompleteEnabled` is probably a better name here

dnachum commented:
Thanks for reviewing this! I'll try to work on getting this to work better for some of the special forms.
